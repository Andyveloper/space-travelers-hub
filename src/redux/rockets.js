import { createSlice } from '@reduxjs/toolkit';

const initialState = {
  rockets: [],
};

const reducer = createSlice({
  name: 'rockets',
  initialState,
  reducers: {
    updateRocket: (state, action) => ({
      ...state,
      rockets: {
        ...state.rockets,
        [action.payload]: {
          ...state.rockets[action.payload],
          reserverd: !state.rockets[action.payload].reserved,
        },
      },
    }),
  },
});

export const { updateRocket } = reducer.actions;
export default reducer;
