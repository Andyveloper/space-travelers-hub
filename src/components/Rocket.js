import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { updateRocket } from '../redux/rockets';

const Rocket = (props) => {
  const dispatch = useDispatch();

  const {
    name, description, image, id,
  } = props;

  return (
    <li className="rocket-container">
      <img className="rocket-container__img" src={image} alt={name} />
      <div className="rocket-info">
        <h2 className="rocket-info__title">{name}</h2>
        <p className="rocket-info__description">{description}</p>
        <button className="rocket-info__button" type="button" onClick={() => dispatch(updateRocket(id))}>Test</button>
      </div>
    </li>
  );
};

Rocket.propTypes = {
  name: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
  id: PropTypes.number.isRequired,
};

export default Rocket;
