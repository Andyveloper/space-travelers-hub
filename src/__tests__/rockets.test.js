import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Rocket from '../components/Rocket';
import getRockets from '../redux/apiRockets';
import rockets from '../redux/rockets';

const renderReservedRockets = (reserved) => {
  const { tree } = render(
    <Provider store={store}>
      <Rocket
        reserved={reserved}
        id={1}
        name="Falcon 1"
        description="Test description"
        image="https://imgur.com/DaCfMsj.jpg"
      />
    </Provider>,
  );
  return tree;
};

describe('Test rocket button and message', () => {
  it('Renders first Rocket', () => {
    expect(renderReservedRockets(true)).toMatchSnapshot();
  });

  it('Shows Reserved message in description and Cancel Reservation button', () => {
    renderReservedRockets(true);
    expect(screen.getByText(/Reserved/i)).toBeInTheDocument();
    expect(screen.getByText(/Cancel Reservation/i)).toBeInTheDocument();
  });

  it('Shows Reserve Rocket button', () => {
    renderReservedRockets(false);
    expect(screen.getByText(/Reserve Rocket/i)).toBeInTheDocument();
  });
});

describe('Test rocket reducers and functions', () => {
  it('Fetches data from API', async () => {
    await store.dispatch(getRockets());
    expect(store.getState().rockets).toBeTruthy();
  });

  it('Returns the initial state', () => {
    expect(rockets.reducer(undefined, { type: undefined })).toEqual(
      { rockets: [] },
    );
  });
});
