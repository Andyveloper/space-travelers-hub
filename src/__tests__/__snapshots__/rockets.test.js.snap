// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test rocket button and message Renders first Rocket 1`] = `undefined`;
